### PR TITLE
fix: `notation plugin install` error messages and tests

### LINKED ISSUE #853: notation plugin command enhancements
_Originally posted by @priteshbandi in https://github.com/notaryproject/notation/pull/827#pullrequestreview-1792218140_

@priteshbandi  Created this issue to track future works for `notation plugin install` commands. Targeting v1.1.0 release.

### PR DESCRIPTION
**Pull request by Two-Hearts**
This PR improves error messages of `notation plugin install` command and updates the related cli specs as well.

This PR also adds zip bomb and zip slip tests.

Resolves #853.

### CODE REVIEW

File: internal/osutil/file.go [hunk 1]
```diff
@@ -157,7 +158,7 @@ func ValidateSHA256Sum(path string, checksum string) error {
 	sha256sum := sha256Hash.Sum(nil)
 	enc := hex.EncodeToString(sha256sum[:])
 	if !strings.EqualFold(enc, checksum) {
-		return fmt.Errorf("plugin checksum does not match user input. Expecting %s", checksum)
+		return fmt.Errorf("plugin sha256sum does not match user input. Expecting %s", checksum)
```
Comment (shizhMSFT) [inline internal/osutil/file.go:161]:
```suggestion
		return fmt.Errorf("plugin SHA-256 checksum does not match user input. Expecting %s", checksum)
```

File: test/e2e/testdata/malicious-plugin/largeFileZip.zip
[the remark this answers is not in the document]
Comment (Two-Hearts) [reply test/e2e/testdata/malicious-plugin/largeFileZip.zip:1]: updated to non sparse files

**Review by shizhMSFT — DISMISSED** [review]
LGTM

File: cmd/notation/plugin/install.go [hunk 2]
```diff
@@ -103,8 +106,8 @@ Example - Install plugin from URL, SHA256 checksum is required:
 		},
 	}
 	opts.LoggingFlagOpts.ApplyFlags(command.Flags())
-	command.Flags().BoolVar(&opts.isFile, "file", false, "install plugin from a file in file system")
-	command.Flags().BoolVar(&opts.isURL, "url", false, "install plugin from an HTTPS URL. The timeout of the download HTTPS request is set to 10 minutes")
+	command.Flags().BoolVar(&opts.isFile, "file", false, "install plugin from a file on file system")
+	command.Flags().BoolVar(&opts.isURL, "url", false, fmt.Sprintf("install plugin from an HTTPS URL. The default plugin download timeout is %s", notationplugin.DownloadPluginFromURLTimeout))
```
Comment (priteshbandi) [inline cmd/notation/plugin/install.go:110]:
```suggestion
	command.Flags().BoolVar(&opts.isURL, "url", false, fmt.Sprintf("install plugin from an HTTPS URL. The plugin download timeout is %s", notationplugin.DownloadPluginFromURLTimeout))
```

This is not a customizable so no default value.

File: specs/commandline/plugin.md [hunk 3]
```diff
@@ -48,11 +48,11 @@ Usage:
...
   -d, --debug              debug mode
-      --file               install plugin from a file in file system
-      --force              force the installation of a plugin
+      --file               install plugin from a file on file system
+      --force              force the installation of the plugin
   -h, --help               help for install
-      --sha256sum string   must match SHA256 of the plugin source
-      --url                install plugin from an HTTPS URL
+      --sha256sum string   must match SHA256 of the plugin source, required when "--url" flag is set
+      --url                install plugin from an HTTPS URL. The default plugin download timeout is 10m0s
```
Comment (priteshbandi) [inline specs/commandline/plugin.md:55]:
```suggestion
      --url                install plugin from an HTTPS URL. The plugin download timeout is 10m0s
```

File: specs/commandline/plugin.md [hunk 4]
```diff
@@ -80,7 +82,7 @@ Aliases:
 
 ### Install a plugin from file system
 
-Install a Notation plugin from file system. Plugin file supports `.zip` and `.tar.gz` format. The checksum validation is optional for this case. 
+Install a Notation plugin from file system. Plugin file supports `.zip`, `.tar.gz`, and single plugin executable file format. The checksum validation is optional for this case.
```
Comment (priteshbandi) [inline specs/commandline/plugin.md:85]:
```suggestion
Install a Notation plugin from the file system. When installing a plugin from a remote location, both `.zip` and `.tar.gz` file formats are supported. Additionally, installing a plugin from an executable file on the host system is also supported. In this scenario, checksum validation is optional.
```

File: test/e2e/suite/plugin/install.go [hunk 5]
```diff
@@ -60,24 +60,38 @@ var _ = Describe("notation plugin install", func() {
 		})
 	})
 
+	It("with content inside zip archive exceeds 256 MiB size limit", func() {
```
Comment (priteshbandi) [inline test/e2e/suite/plugin/install.go:63]: nit: Please add test for zipbomb and zip slip. The zips
1. Zipbomb sample: https://www.bamsoftware.com/hacks/zipbomb/
2. ZipSlip sample: `zip zipslip.zip ../../../../../../../../tmp/evil.sh`

File: specs/commandline/plugin.md (hunk 4 above)
Comment (Two-Hearts) [reply specs/commandline/plugin.md:85]: @priteshbandi This paragraph is for `installing from file system` only. Updated both `install from file system` and `install from URL` based on your comments.

File: test/e2e/suite/plugin/install.go (hunk 5 above)
Comment (Two-Hearts) [reply test/e2e/suite/plugin/install.go]: Tests added.

File: test/e2e/suite/plugin/install.go [hunk 6]
```diff
@@ -60,24 +60,52 @@ var _ = Describe("notation plugin install", func() {
...
+	It("with zip bomb single file exceeds 256 MiB size limit in zip format", func() {
+		Host(nil, func(notation *utils.ExecOpts, _ *Artifact, vhost *utils.VirtualHost) {
+			notation.ExpectFailure().Exec("plugin", "install", "--file", NotationE2EMaliciousPluginArchivePath+"/large_file_zip.zip", "-v").
+				MatchErrContent("Error: plugin installation failed: total file size reached the 256 MiB size limit\n")
+		})
+	})
+
+	It("with zip bomb single file exceeds 256 MiB size limit in tar.gz format", func() {
+		Host(nil, func(notation *utils.ExecOpts, _ *Artifact, vhost *utils.VirtualHost) {
+			notation.ExpectFailure().Exec("plugin", "install", "--file", NotationE2EMaliciousPluginArchivePath+"/large_file_tarGz.tar.gz", "-v").
```
Comment (JeyJeyGao) [inline test/e2e/suite/plugin/install.go:72]:
```suggestion
			notation.ExpectFailure().Exec("plugin", "install", "--file", filepath.Join(NotationE2EMaliciousPluginArchivePath, "large_file_tarGz.tar.gz"), "-v").
```
For cross-platform E2E support in the future.

File: cmd/notation/plugin/install.go [hunk 7]
```diff
@@ -185,12 +185,21 @@ func installPlugin(ctx context.Context, inputPath string, inputChecksum string,
 			return err
 		}
 		defer rc.Close()
+		// check for '..' in file name to avoid zip slip vulnerability
+		for _, f := range rc.File {
+			if strings.Contains(f.Name, "..") {
+				return fmt.Errorf("file name in zip cannot contain '..', but found %q", f.Name)
+			}
+		}
```
Comment (shizhMSFT) [inline cmd/notation/plugin/install.go:193]: What if a file name is `foo..bar.txt`?

Comment (Two-Hearts) [reply cmd/notation/plugin/install.go:193]: We do not accept such file either.

Comment (Two-Hearts) [reply cmd/notation/plugin/install.go:193]: This was recommended by the `github-advanced-security`: https://github.com/notaryproject/notation/security/code-scanning/44

**Review by shizhMSFT — APPROVED** [review]
LGTM

File: cmd/notation/plugin/install.go (hunk 7 above)
Comment (shizhMSFT) [reply cmd/notation/plugin/install.go:193]: It is over-kill but I guess we can fix it later when we got a real file with `..` in it.

Comment (Two-Hearts) [reply cmd/notation/plugin/install.go:193]: I see. Resolving this comment for now.

**Review by priteshbandi — APPROVED** [review]
LGTM

**Pull request merged**